File: app/src/main/java/com/abhijeetpratap/greens/data/ProductDatabase.java
package com.abhijeetpratap.greens.data;

import android.content.Context;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;

@androidx.room.Database(entities = {Product.class}, version = 2)
@TypeConverters(CategoryConverter.class)
public abstract class ProductDatabase extends RoomDatabase {
    private static ProductDatabase INSTANCE;

    public static ProductDatabase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (ProductDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                            ProductDatabase.class, "items_database")
                            .fallbackToDestructiveMigration()
                            .build();
                }
            }
        }
        return INSTANCE;
    }

    public abstract ProductDao productDao();
}

File: app/src/main/java/com/abhijeetpratap/greens/data/ProductDatabase.kt
package com.abhijeetpratap.greens.data

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase

@Database(entities = [Product::class], version = 1)
abstract class ProductDatabase : RoomDatabase() {
    abstract fun productDao(): ProductDao

    companion object {
        @Volatile
        private var instance: ProductDatabase? = null
        private val DATABASE_NAME = "all_products"

        fun getInstance(context: Context): ProductDatabase {
            return instance ?: synchronized(this) {
                instance ?: builDatabase(context).also {
                    instance = it
                }
            }
        }

        private fun builDatabase(context: Context): ProductDatabase {
            return Room.databaseBuilder(
                context,
                ProductDatabase::class.java, DATABASE_NAME
            ).build()
        }
    }
}
